### ISSUE TITLE
Attacking causes Hitboxie to fall through platforms

### ISSUE THREAD

**Issue by eaglgenes101** [digiholic/universalSmashSystem#8]
Hitboxie's forward tilt, forward smash, and neutral aerial cause Hitboxie to fall through platforms, apparently because Hitboxie's bounding box bottom changes in the middle. Could the engine be modified so that regardless of action, the "feetbox" of all fighters don't change position relative to the fighter?

**Comment by digiholic** [comment]
I've been aware of the platform clipping, but couldn't come up with a solution. A footbox is a good idea, I'll work on that next.

**Comment by meffij** [comment]
Maybe just a flag that can be set for grounded moves that prevents the fighter from moving vertically relative to the platform they're standing on as long as the move continues. This would prevent fsmash, etc. from clipping through the platform, while allowing falling through the platform if the move is interrupted. This would also stick the fighter to the platform if it moves, while still allowing horizontal sliding i.e. wavedash ftilt and the like.

**Comment by eaglgenes101** [comment]
I wouldn't suggest that solution. This problem also happens with aerials, and it's going to mess up a lot of aerial-platform play if certain attacks that cause the hitbox bottom to move cause the fighter to fall through platforms instead of landing on them. 

Speaking of, there's probably a cleaner solution: Determine hitbox placements before movements, and check to see if there's a platform below before positioning the fighter, and do anti-clipping if this would happen. I'll try to pull together a pull request for this.

**Comment by meffij** [comment]
In Project M, the ground collision point of a character is defined in the animation the character is in. So the only way to fall through a platform from the air is to be in an airstate that allows for falling through the platform, and to be holding down. To make the engine as versatile as possible, it makes sense to me that the individual moves of a fighter should be able to set platform fallthrough rules. For example, maybe someone wants to make a move that cannot land on a platform, the flag is never set. Or a move that falls through the platform unless the player is holding A, set the flag based on input. Stuff like that is impossible if you make the decision in the engine.

Additionally, my solution allows for less complicated ground move code. We want to make creating and tweaking fighters as painless as possible, and having simple flags for common use cases can help that happen. I believe that most ground moves are intended to stay on the ground, and having a flag that captures those semantics seems reasonable.

I think we should be pursuing solutions that define engine behavior in individual fighter files as much as possible. By minimizing the assumptions we make about the behaviors of fighters, we can allow our users to get more creative and feel less constrained by the engine. Users will likely prefer tweaking an existing fighter to writing one from scratch anyway, so having simple flags that capture bundles of common behavior for moves, animations, etc. will allow users to get what they imagine without rewriting code all over the place. This will help people make their new move work like some existing move by compartmentalizing common functionality in an easily-identifiable line of code in the move.

TL;DR Aerials should generally fall through the platform if they start when the fighter's foot is below the platform, putting fighter design decisions in the engine makes letting users customize fighter behavior harder or impossible, and it will be easier for people to create fighters if we have simple ways to assign common behavior to aspects of the fighter, because that's how people will be thinking when they're building fighters.

**Comment by ghost** [comment]
Our current plan is to have a cross of colliders that mimic the diamond shaped ECBs in Melee, Digi can probably explain better though.

**Comment by eaglgenes101** [comment]
@meffij I get what your saying, but I doubt that most designers want their fighters to fall through platforms while doing aerials. Fix the core issue first, then we can get to that snazzy ground stuff you want.

**Comment by meffij** [comment]
What do you mean by fall through platforms during aerials? In the current
build, Hitboxie lands on the platform when he nairs.

I think the earlier we think about these sorts of things, the easier they
will be to fix. I think it's a good idea to build systems that provide
functionality like ECBs, but I think the more we write them into the engine
the harder it will be to deviate from the standard. If the default fighters
process collisions with a standard system, that's great, but if that
standard system is part of the engine then people will write really hacky
horrible stuff to try to get around it, and ultimately will end up with a
buggy unfinished character and lots of frustration.

Maybe I'm misunderstanding your intention, and you're saying you're going
to move the improved code into abstractFighter or something. If so, sorry.

I think the closer the engine is to a library of useful functions and an
event loop manager, the better. The more core functionality that belongs to
the stages and fighters, the better. And making stages and fighters heavier
isn't creating work, since we're using a very featureful base abstract
class for both. It'll be easy to make sure users don't have to think about
anything they don't want to modify from default, and to offer several
standard 'presets' for stuff like stage cameras, fighter input handling,
etc. So if you want to make a fighter with standard tilts, aerials,
smashes, dodges, shield, etc. then all you'll have to do is implement the
moves and provide sprites.

**Comment by eaglgenes101** [comment]
Not making it up. Hitboxie can fall through platforms if it jumps and it is doing a nair at a time it contacts a platform: https://youtu.be/IApEqkU7-nc

And yes, this is supposed to file under the AbstractFighter. Sorry that I didn't make it clear.

**Comment by meffij** [comment]
I see, the sprite transition during the move makes it go through the platform. I was thinking something along the lines of
--be above platform
-- start move near enough to the platform that first frame of move puts bottom collision box below platform
-- do not land on platform
which didn't seem like a problem to me, hence my confusion. I appreciate you going to the trouble to make a video to help me understand the issue. Thanks!

What are your thoughts on a developer chat? Like maybe Discord or something? I don't really have a suggestion, other than we should pick something that's not terrible and start using it. I think it's a better idea to have something like that than for me to keep clogging up our issue tracker with clarification requests.

Thanks!

**Comment by digiholic** [comment]
So, right now, the platform collision comes from a pixel-perfect rect collision. If you're not colliding with something, but you then change sprites into something that does pass through it, the game thinks you've come up from the platform and hit the peak of your jump not quite through the platform yet (since your last frame wasn't actually above the platform) and drops you. This happens during forward smash because you're gravity affected during the lunge, but not at the end when you switch back, falling through. If your nair is timed right so that your sprite sort of slices through the platform you will fall through.

What we're doing to fix this is adding in an Environment Collision Cross. Two thin rects that span the width and height of the sprite, that will be generated automatically, but can also be manually set in the actions. This means that for the most part, you can easily allow the game to set the collision mask, but for some problem animations, do a little bit of extra work to define the cross manually.

For more information, here's a video on Melee's ECB that I'm using for reference purposes:
[https://www.youtube.com/watch?v=TtiDd5cQWJc](https://www.youtube.com/watch?v=TtiDd5cQWJc)

As for a developer chat, we've got a Slack team, you're welcome to join if you'd like. If you want to give me your email, I'll add you.

**Comment by eaglgenes101** [comment]
I created a pull request that fixes this problem for aerials. It checks in front of itself to not be caught off-guard by platforms, which helps tremendously. There's probably a way to break it, but it's much more solid than it used to be. 

I don't know what's going on for the grounded moves, on the other hand.

**Comment by meffij** [comment]
Is there a good way to send you my email without making it public?

**Comment by meffij** [comment]
This issue is still present after the ECB rewrite. I replicated the issue by charging a forward smash on the side platform of arena_moving_platform. Interestingly, charging a forward smash on the moving platform did not cause Hitboxie to fall through.

**Comment by eaglgenes101** [comment]
Alright, I submitted a new pull request to handle certain cases where hitboxie nair would occasionally fall through moving platforms. 

I'm still trying to figure out what's going on with forward tilt and forward smash.

**Comment by digiholic** [comment]
I haven't even done the ECB rewrite yet. It should hopefully be done in a few days, which will fix the smash and tilt issues, as well as clipping through during hitstun.

**Comment by eaglgenes101** [comment]
I believe I've finally resolved this problem for good. If you find a problem with the collision handling, open a new issue so I know to eat my own words.
